circuit/algorithms/src/bhp/hasher/hash_uncompressed.rs

```diff
@@ -0,0 +1,232 @@
+                // Pad the input to a multiple of `BHP_CHUNK_SIZE` for hashing.
+                if input.len() % BHP_CHUNK_SIZE != 0 {
+                    let padding = BHP_CHUNK_SIZE - (input.len() % BHP_CHUNK_SIZE);
+                    input.resize(input.len() + padding, Boolean::constant(false));
+                    assert_eq!(input.len() % BHP_CHUNK_SIZE, 0, "Input must be a multiple of {BHP_CHUNK_SIZE}");
+                }
+            }
+            false => E::halt(format!("Inputs to this BHP cannot exceed {} bits", Self::MAX_BITS)),
+        }
+
+        // Declare the 1/2 constant field element.
+        let one_half = Field::constant(E::BaseField::half());
+
+        // Declare the constant coefficients A and B for the Montgomery curve.
+        let coeff_a = Field::constant(<E::AffineParameters as TwistedEdwardsParameters>::MontgomeryParameters::COEFF_A);
+        let coeff_b = Field::constant(<E::AffineParameters as TwistedEdwardsParameters>::MontgomeryParameters::COEFF_B);
+
+        // Implements the incomplete addition formulae of two Montgomery curve points.
+        let montgomery_add = |(this_x, this_y): (&Field<E>, &Field<E>), (that_x, that_y): (&Field<E>, &Field<E>)| {
+            // Construct `lambda` as a witness defined as:
+            // `lambda := (that_y - this_y) / (that_x - this_x)`
+            let lambda: Field<E> = witness!(|this_x, this_y, that_x, that_y| (that_y - this_y) / (that_x - this_x));
+
+            // Ensure `lambda` is correct by enforcing:
+            // `lambda * (that_x - this_x) == (that_y - this_y)`
+            E::enforce(|| (&lambda, that_x - this_x, that_y - this_y));
+
+            // Construct `sum_x` as a witness defined as:
+            // `sum_x := (B * lambda^2) - A - this_x - that_x`
+            let sum_x: Field<E> = witness!(|lambda, that_x, this_x, coeff_a, coeff_b| {
+                coeff_b * lambda.square() - coeff_a - this_x - that_x
+            });
+
+            // Ensure `sum_x` is correct by enforcing:
+            // `(B * lambda) * lambda == (A + this_x + that_x + sum_x)`
+            E::enforce(|| (&coeff_b * &lambda, &lambda, &coeff_a + this_x + that_x + &sum_x));
+
+            // Construct `sum_y` as a witness defined as:
+            // `sum_y := -(this_y + (lambda * (this_x - sum_x)))`
+            let sum_y: Field<E> = witness!(|lambda, sum_x, this_x, this_y| -(this_y + (lambda * (sum_x - this_x))));
+
+            // Ensure `sum_y` is correct by enforcing:
+            // `(lambda * (this_x - sum_x)) == (this_y + sum_y)`
+            E::enforce(|| (&lambda, this_x - &sum_x, this_y + &sum_y));
+
+            (sum_x, sum_y)
+        };
+
+        // Compute sum of h_i^{sum of (1-2*c_{i,j,2})*(1+c_{i,j,0}+2*c_{i,j,1})*2^{4*(j-1)} for all j in segment}
+        // for all i. Described in section 5.4.1.7 in the Zcash protocol specification.
+        //
+        // Note: `.zip()` is used here (as opposed to `.zip_eq()`) as the input can be less than
```

Comment:
Is this comment still true? Now that we’re padding everywhere, we might be able to use `zip_eq`

Reply:
Yes, for the final round of BHP, it is still the case.